File: esp_system/inject/src/com/esp/EspFrame.java
package com.esp;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.List;

/**
 * ESP 数据帧 (TVEF v2 协议) — 与 tv_reader 输出格式一致。
 * 由 Kotlin 版 EspFrame.kt 移植 (注入 dex 无 Kotlin 运行时)。
 */
public class EspFrame {

    public static class Skill {
        public int spellId;
        public float cooldownRemaining;
        public float cooldownTotal;
        public boolean ready;
    }

    public static class Actor {
        public int type;
        public float x, z;
        public boolean ally;
        public int hp, maxHp;
        public boolean visible;
        public int nameId;
        public int level;
        public float ultimateCooldown, ultimateTotal;
        public boolean ultimateReady;
        public List<Skill> spells;
        public float facingAngle;
        public float speed;

        public boolean isDrawable() { return type >= 1 && type <= 50; }
        public boolean isHero()     { return type >= 1 && type <= 50; }
        public boolean isTower()    { return type >= 60 && type <= 100; }
        public boolean isMinion()   { return type >= 100 && type <= 200; }
        public boolean isMonster()  { return type >= 200 && type <= 300; }
        public boolean isDragon()   { return type >= 300 && type <= 350; }
        public boolean isBaron()    { return type >= 350 && type <= 400; }

        public float ultimateRatio() {
            if (ultimateTotal > 0f) {
                return clamp(1f - ultimateCooldown / ultimateTotal, 0f, 1f);
            }
            return 1f;
        }

        public float hpRatio() {
            return maxHp > 0 ? (float) hp / (float) maxHp : 0f;
        }
    }

    public static class Timer {
        public int id;
        public float respawnSeconds;
        public float maxSeconds;
        public boolean active;
        public String label;

        public float ratio() {
            if (maxSeconds > 0f) {
                return clamp(1f - respawnSeconds / maxSeconds, 0f, 1f);
            }
            return 0f;
        }
    }

    public long frameId;
    public List<Actor> actors = new ArrayList<>();
    public List<Timer> timers = new ArrayList<>();
    public float gameTime, selfX, selfZ;
    public String status = "ok";

    private static float clamp(float v, float lo, float hi) {
        return v < lo ? lo : (v > hi ? hi : v);
    }

    public static EspFrame parse(byte[] data) {
        ByteBuffer buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
        byte[] magic = new byte[4];
        buf.get(magic);
        if (!new String(magic).equals("TVEF")) {
            throw new IllegalArgumentException("Bad magic");
        }
        int version = buf.get() & 0xFF;
        if (version != 2) {
            throw new IllegalArgumentException("Bad version " + version);
        }

        EspFrame f = new EspFrame();
        f.frameId = buf.getInt() & 0xFFFFFFFFL;
        f.gameTime = buf.getFloat();
        f.selfX = buf.getFloat();
        f.selfZ = buf.getFloat();

        int actorCount = buf.get() & 0xFF;
        for (int i = 0; i < actorCount; i++) {
            Actor a = new Actor();
            a.type = buf.getInt();
            a.x = buf.getFloat();
            a.z = buf.getFloat();
            a.ally = buf.get() != 0;
            a.hp = buf.getInt();
            a.maxHp = buf.getInt();
            a.visible = buf.get() != 0;
            a.nameId = buf.getInt();
            a.level = buf.get() & 0xFF;
            a.ultimateCooldown = buf.getFloat();
            a.ultimateTotal = buf.getFloat();
            int spellCount = buf.get() & 0xFF;
            a.spells = new ArrayList<>(spellCount);
            for (int s = 0; s < spellCount; s++) {
                Skill sk = new Skill();
                sk.spellId = buf.getInt();
                sk.cooldownRemaining = buf.getFloat();
                sk.cooldownTotal = buf.getFloat();
                sk.ready = buf.get() != 0;
                a.spells.add(sk);
            }
            a.facingAngle = buf.getFloat();
            a.speed = buf.getFloat();
            a.ultimateReady = a.ultimateCooldown <= 0f;
            f.actors.add(a);
        }

        int timerCount = buf.get() & 0xFF;
        for (int i = 0; i < timerCount; i++) {
            Timer t = new Timer();
            t.id = buf.getInt();
            t.respawnSeconds = buf.getFloat();
            t.maxSeconds = buf.getFloat();
            t.active = buf.get() != 0;
            byte[] lb = new byte[12];
            buf.get(lb);
            String label = new String(lb);
            int end = label.indexOf('\0');
            t.label = end >= 0 ? label.substring(0, end) : label;
            f.timers.add(t);
        }
        return f;
    }
}

File: esp_system/android/app/src/main/java/com/esp/EspFrame.kt
package com.esp

import java.nio.ByteBuffer
import java.nio.ByteOrder

data class EspSummonerSkill(
    val spellId: Int,
    val cooldownRemaining: Float,
    val cooldownTotal: Float,
    val ready: Boolean
)

data class EspActor(
    val type: Int,
    val x: Float,
    val z: Float,
    val y: Float,
    val ally: Boolean,
    val hp: Int,
    val maxHp: Int,
    val visible: Boolean,
    val nameId: Int,
    val level: Int,
    val ultimateCooldown: Float,
    val ultimateTotal: Float,
    val ultimateReady: Boolean,
    val summonerSpells: List<EspSummonerSkill>,
    val facingAngle: Float,
    val speed: Float
) {
    val drawable: Boolean get() = type in 1..50
    val isHero: Boolean get() = type in 1..50
    val isTower: Boolean get() = type in 60..100
    val isMinion: Boolean get() = type in 100..200
    val isMonster: Boolean get() = type in 200..300
    val isDragon: Boolean get() = type in 300..350
    val isBaron: Boolean get() = type in 350..400
    val isPlayer: Boolean get() = isHero

    val ultimateRatio: Float
        get() = if (ultimateTotal > 0f) (1f - ultimateCooldown / ultimateTotal).coerceIn(0f, 1f) else 1f

    val hpRatio: Float
        get() = if (maxHp > 0) hp.toFloat() / maxHp else 0f
}

data class EspGlobalTimer(
    val id: Int,
    val respawnSeconds: Float,
    val maxSeconds: Float,
    val active: Boolean,
    val label: String
) {
    val ratio: Float
        get() = if (maxSeconds > 0f) (1f - respawnSeconds / maxSeconds).coerceIn(0f, 1f) else 0f
}

data class EspFrame(
    val frameId: UInt,
    val actors: List<EspActor>,
    val timers: List<EspGlobalTimer>,
    val gameTime: Float,
    val selfX: Float,
    val selfZ: Float,
    val selfY: Float,
    val status: String
) {
    companion object {
        private const val MAGIC = "TVEF"

        private const val SKILL_SIZE = 13
        private const val TIMER_SIZE = 19

        fun parse(data: ByteArray): EspFrame {
            val buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
            val magic = ByteArray(4).also { buf.get(it) }
            if (String(magic) != MAGIC) throw IllegalArgumentException("Bad magic")
            val version = buf.get().toInt()
            if (version !in 2..3) throw IllegalArgumentException("Bad version $version")
            val frameId = buf.int.toUInt()
            val gameTime = buf.float
            val selfX = buf.float
            val selfZ = buf.float
            // v3 新增: 自身高度 (用于屏幕投影)
            val selfY = if (version >= 3) buf.float else 0f

            val actorCount = buf.get().toInt() and 0xFF
            val actors = (0 until actorCount).map {
                val type = buf.int
                val x = buf.float
                val z = buf.float
                // v3 新增: actor 高度
                val y = if (version >= 3) buf.float else 0f
                val ally = buf.get() != 0.toByte()
                val hp = buf.int
                val maxHp = buf.int
                val visible = buf.get() != 0.toByte()
                val nameId = buf.int
                val level = buf.get().toInt() and 0xFF
                val ultCd = buf.float
                val ultTotal = buf.float
                val spellCount = buf.get().toInt() and 0xFF
                val spells = (0 until spellCount).map {
                    EspSummonerSkill(
                        spellId = buf.int,
                        cooldownRemaining = buf.float,
                        cooldownTotal = buf.float,
                        ready = buf.get() != 0.toByte()
                    )
                }
                val facing = buf.float
                val speed = buf.float
                EspActor(
                    type = type, x = x, z = z, y = y, ally = ally,
                    hp = hp, maxHp = maxHp, visible = visible,
                    nameId = nameId, level = level,
                    ultimateCooldown = ultCd, ultimateTotal = ultTotal,
                    ultimateReady = ultCd <= 0f,
                    summonerSpells = spells,
                    facingAngle = facing, speed = speed
                )
            }

            val timerCount = buf.get().toInt() and 0xFF
            val timers = (0 until timerCount).map {
                val id = buf.int
                val respawn = buf.float
                val max = buf.float
                val active = buf.get() != 0.toByte()
                val labelBytes = ByteArray(12).also { buf.get(it) }
                val label = String(labelBytes).trimEnd('\u0000')
                EspGlobalTimer(id, respawn, max, active, label)
            }

            return EspFrame(frameId, actors, timers, gameTime, selfX, selfZ, selfY, "ok")
        }
    }
}

data class EspStatus(
    val connected: Boolean,
    val frame: EspFrame? = null,
    val msg: String = ""
)
